### PR TITLE
Speed up list_available

### PR DESCRIPTION
**Pull request by ericpre**
As discussed in https://github.com/mamba-org/gator/issues/132 and https://github.com/mamba-org/mamba/issues/835, use `mamba repoquery search "*" --json` in favour of `mamba search --json` (or `conda search --json`), since the later is significantly slower than `mamba repoquery`.

### CODE REVIEW

**Comment by ericpre** [comment]
The test seem to be playing up, no sure if this is related to PR, because running the test suite on the master isn't working - see for example https://github.com/ericpre/gator/actions/runs/722378324.

**Comment by wolfv** [comment]
Cool, thanks for working on this!

**Comment by fcollonval** [comment]
Hey @ericpre 

thanks for the PR. I carried out some performance tests (absolute figure does not matter much here):

image

Your solution will definitely improve experienced for `mamba` users.

@wolfv I'm not familiar with the mamba details. Does `mamba repoquery search` update the packages cache if it is expired?

File: mamba_gator/envmanager.py [hunk 1]
```diff
@@ -596,6 +599,25 @@ async def list_available(self) -> Dict[str, List[Dict[str, str]]]:
...
+            of "conda search --json".
+            """
+
+            data_ = {}
+            for entry in data['result']['pkgs']:
+                name = entry.get('name')
+                if name in data_.keys():
+                    data_[name].append(entry)
+                else:
+                    data_[name] = [entry]
```
Comment (fcollonval) [inline mamba_gator/envmanager.py:614]: You could simplify this like this:

```suggestion
            data_ = collections.defaultdict(lambda : [])
            for entry in data['result']['pkgs']:
                name = entry.get('name')
                if name is not None:
                    data_[name].append(entry)
```

Under the hood the default dict will create an empty list for data_[name] if name key does not exist.

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @ericpre for this PR. I left a few improvements.

Could you also add a test for it?

For that the best would be to check the call of `f` in 

https://github.com/ericpre/gator/blob/3b9c1845f2fed9c8e2997ecba9dabcc560ad29d0/mamba_gator/tests/test_api.py#L794

You could test the call with something like:

```
f.assert_called_with(...)
```

after line 975. Note the call arguments will depend on the manager.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -2,7 +2,7 @@ name: Test
 
 on:
   push:
-    branches: [master]
+    branches: "*"
```
[the remark this answers is not in the document]
Comment (ericpre) [reply .github/workflows/build.yml]: Yes, I have removed this commit.

File: mamba_gator/envmanager.py [hunk 3]
```diff
@@ -585,7 +585,10 @@ async def list_available(self) -> Dict[str, List[Dict[str, str]]]:
                 "with_description": bool  # Whether we succeed in get some channeldata.json files
             }
         """
-        ans = await self._execute(self.manager, "search", "--json")
+        if Path(self.manager).stem != "mamba":
```
[the remark this answers is not in the document]
Comment (ericpre) [reply mamba_gator/envmanager.py]: Done.

File: mamba_gator/envmanager.py (hunk 1 above)
Comment (ericpre) [reply mamba_gator/envmanager.py]: Indeed, this is more clear, done!

**Comment by ericpre** [comment]
> Thanks @ericpre for this PR. I left a few improvements.
> 
> Could you also add a test for it?
> 
> For that the best would be to check the call of `f` in
> 
> https://github.com/ericpre/gator/blob/3b9c1845f2fed9c8e2997ecba9dabcc560ad29d0/mamba_gator/tests/test_api.py#L794
> 
> You could test the call with something like:
> 
> ```
> f.assert_called_with(...)
> ```
> 
> after line 975. Note the call arguments will depend on the manager.

Sorry, I am not sure that I understand, what should be tested exactly? Do you want to have the ouptut of `self._execute(self.manager, "repoquery", "search", "*", "--json")` tested?
`list_available` is already tested and the changes of this PR should be covered by these tests?

**Comment by fcollonval** [comment]
> Sorry, I am not sure that I understand, what should be tested exactly? Do you want to have the ouptut of `self._execute(self.manager, "repoquery", "search", "*", "--json")` tested?
> `list_available` is already tested and the changes of this PR should be covered by these tests?

As you said, the output is already tested. I would like only to add a test to be sure the proper command is invoked depending on the manager; i.e. if conda, `_execute` should be called with ["conda", "search", "--json"] - if mamba, it should be called with [manager, "repoquery", "search", "*", "--json"].

Does it clarify the intention?

**Comment by ericpre** [comment]
Yes, it makes sense, thanks!
It seems that the tests need to be adapted to pass the changes of this PR and since I don't understand how these tests works, it would take me quite a bit time to get it to work and I will able to do this here. Can you please take over the PR?

**Review by fcollonval — APPROVED** [review]

**Comment by fcollonval** [comment]
Thanks @ericpre I added the tests.

**Comment by ericpre** [comment]
Thanks for sorting out the test and finishing the PR!

**Pull request merged**